src/server/builder.ts

```diff
@@ -350,9 +352,15 @@ namespace ts.server {
                         }
                     }
                 }
-                fileNameSet[processingFileInfo.scriptInfo.fileName] = true;
+                fileNameSet[processingFileInfo.scriptInfo.fileName] = processingFileInfo.scriptInfo;
             }
-            return Object.keys(fileNameSet);
+            const result: string[] = [];
+            for (const fileName in fileNameSet) {

```

Comment:
make a function `shouldEmitFile`:
```ts
function shouldEmitFile(scriptInfo: ScriptInfo) {
    return !scriptInfo.hasMixedContent;
}
```
then you can use it here and in calls to `filter` above?